// @umijs/preset-dumi/lib/themes/default/context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = _react().default.createContext({
  /**
   * current root path (with locale prefix)
   */
  rootPath: '/',

  /**
   * current slug
   */
  slug: '',

  /**
   * current locale name
   */
  locale: '',

  /**
   * all available locales
   */
  locales: [],

  /**
   * mode
   */
  mode: 'doc',

  /**
   * logo URL
   */
  logo: '',

  /**
   * title
   */
  title: '',

  /**
   * desc
   */
  desc: '',

  /**
   * repository url
   */
  repoUrl: '',

  /**
   * nav list for current locale
   */
  navs: [],

  /**
   * menu list for current locale with current path
   */
  menus: [],

  /**
   * meta of current route
   */
  routeMeta: {}
});

exports.default = _default;
