// @umijs/preset-dumi/lib/transformer/remark/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _unified() {
  const data = _interopRequireDefault(require("unified"));

  _unified = function _unified() {
    return data;
  };

  return data;
}

function _remarkFrontmatter() {
  const data = _interopRequireDefault(require("remark-frontmatter"));

  _remarkFrontmatter = function _remarkFrontmatter() {
    return data;
  };

  return data;
}

function _rehypeStringify() {
  const data = _interopRequireDefault(require("rehype-stringify"));

  _rehypeStringify = function _rehypeStringify() {
    return data;
  };

  return data;
}

function _rehypeSlug() {
  const data = _interopRequireDefault(require("rehype-slug"));

  _rehypeSlug = function _rehypeSlug() {
    return data;
  };

  return data;
}

function _rehypeAutolinkHeadings() {
  const data = _interopRequireDefault(require("rehype-autolink-headings"));

  _rehypeAutolinkHeadings = function _rehypeAutolinkHeadings() {
    return data;
  };

  return data;
}

function _rehypeRemoveComments() {
  const data = _interopRequireDefault(require("rehype-remove-comments"));

  _rehypeRemoveComments = function _rehypeRemoveComments() {
    return data;
  };

  return data;
}

function _remarkMath() {
  const data = _interopRequireDefault(require("remark-math"));

  _remarkMath = function _remarkMath() {
    return data;
  };

  return data;
}

function _rehypeKatex() {
  const data = _interopRequireDefault(require("rehype-katex"));

  _rehypeKatex = function _rehypeKatex() {
    return data;
  };

  return data;
}

var _parse = _interopRequireDefault(require("./parse"));

var _rehype = _interopRequireDefault(require("./rehype"));

var _yaml = _interopRequireDefault(require("./yaml"));

var _header = _interopRequireDefault(require("./header"));

var _externalDemo = _interopRequireDefault(require("./externalDemo"));

var _link = _interopRequireDefault(require("./link"));

var _img = _interopRequireDefault(require("./img"));

var _previewer = _interopRequireDefault(require("./previewer"));

var _jsx = _interopRequireDefault(require("./jsx"));

var _isolation = _interopRequireDefault(require("./isolation"));

var _sourceCode = _interopRequireDefault(require("./sourceCode"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * strategy mapping for different transform strategy
 */
const PLUGIN_STRATEGIES = {
  default: [[_remarkFrontmatter().default], [_yaml.default], [_externalDemo.default], [_remarkMath().default], [_rehype.default], [_rehypeKatex().default], [_rehypeStringify().default, {
    allowDangerousHTML: true,
    closeSelfClosing: true
  }], [_rehypeSlug().default], [_rehypeAutolinkHeadings().default], [_header.default], [_link.default], [_img.default], [_rehypeRemoveComments().default, {
    removeConditional: true
  }], [_sourceCode.default], [_previewer.default], [_jsx.default], [_isolation.default]],
  data: [[_remarkFrontmatter().default], [_yaml.default], [_externalDemo.default], [_rehype.default], [_rehypeStringify().default], [_rehypeSlug().default], [_rehypeAutolinkHeadings().default], [_header.default]]
};

var _default = (raw, opts) => {
  const processor = (0, _unified().default)().use(_parse.default, {
    strategy: opts.strategy || 'default',
    langs: opts.previewLangs
  }).data('fileAbsPath', opts.fileAbsPath); // apply plugins through strategy

  PLUGIN_STRATEGIES[opts.strategy].forEach(plugin => {
    processor.use(...plugin);
  });
  return processor.processSync(raw);
};

exports.default = _default;
